### PR TITLE
Duplicate read filter for cell-free DNA data with UMI

### PR DESCRIPTION
**Pull request by takutosato**
@davidbenjamin, I'm going to ask @fleharty to review this one. @fleharty, I'm going to add a test for the change I made in `GATKProtectedVariantContextUtils.java::getAttributeAsInt` on Tuesday, but that shouldn't affect the rest of the code.

### CODE REVIEW

**Comment by droazen** [comment]
@takutosato Migration instructions for this branch: https://github.com/broadinstitute/gatk/wiki/Migrating-branches-from-gatk-protected-to-gatk

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        final int numRefReads = 10;
+        final int numReads = numAltReads + numRefReads;
+
+        final Map<String, List<GATKRead>> readMap = new LinkedHashMap<>();
+        final List<GATKRead> reads = new ArrayList<>(numReads);
+        final byte[] quals = new byte[readLength];
+        Arrays.fill(quals, baseq);
+
+        for (int i = 0; i < numAltReads; i++) {
+            final int startPosition = shiftReads ? alignmentStart + (i % shiftBy.get()) : alignmentStart;
```
Comment (fleharty) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java]: Could you change the name of shiftBy to something like, nUniqueStartStopPositions?
Also, could you add a comment here explaining how you are creating different unique start stop positions by shifting using the mod?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * If evidence for an alternate allele came from a few unique reads, we will filter the variant.
+ * Mutect2FilteringEngine::applyDuplicatedAltReadFilter is the accompanying filter.
+ */
+public class UniqueAltReadCount extends GenotypeAnnotation implements StandardSomaticAnnotation {
+    public static final String UNIQUE_ALT_READ_SET_COUNT_KEY = "UNIQ_ALT_READ_COUNT";
+
+    @Override
+    public List<String> getKeyNames() {
+        return Arrays.asList(UNIQUE_ALT_READ_SET_COUNT_KEY);
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: Using `Collections.singletonList()` is slightly more efficient as long as the result doesn't need to be modified, which is true here and and for `getDescription()`

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                         final GenotypeBuilder gb,
+                         final ReadLikelihoods<Allele> likelihoods) {
+        if (g.isHomRef()){
+            // skip the normal sample
+            return;
+        }
+
+        final Allele altAllele = vc.getAlternateAllele(0); // assume single-allelic
+        final String tumorSampleName = g.getSampleName();
+        Collection<ReadLikelihoods<Allele>.BestAllele> tumorBestAlleles = likelihoods.bestAlleles(tumorSampleName);
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: Need a space after the comma

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java [hunk 4]
```diff
@@ -190,9 +191,28 @@ private void applyEventDistanceFilters(final VariantContext vc, final Collection
         }
     }
 
+    /***
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import htsjdk.variant.vcf.VCFHeaderLineType;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.broadinstitute.hellbender.engine.ReferenceContext;
+import org.broadinstitute.hellbender.utils.genotyper.ReadLikelihoods;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This annotation counts up the number of ALT reads that have a unique start position and fragment length.
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: Reworded the sentence to make it clearer

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+ * We have seen that the majority of false positives in low allele fraction, cell-free DNA have the profile where
+ * the evidence for alternate allele comes solely from one, two, or three sets of PCR-duplicates.
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ * Reads in such a set have the same read-start and mate-end position (i.e. they come from the same insert),
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: good call

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+ * but they have different unique molecular identifiers (UMIs) so UMI-aware Mark Duplicates considers them unique pieces of evidence.
+ * Although these reads have different UMIs, we suspect that they really are PCR-duplicates, for two reasons:
+ * 1) these sites are false positives and 2) with the library we used, which was not TSCA or NEB,
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: So TSCA and NEB aren't considered hybrid selection?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+ * it's highly unlikely that we get multiple fragments with identical start and end positions.
+ * The true cause, we suspect, is a barcode swap, which is the process in which two fragments with the same barcode swap UMIs.
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: yupppp

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * If evidence for an alternate allele came from a few unique reads, we will filter the variant.
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: reworded

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    public static final String UNIQUE_ALT_READ_SET_COUNT_KEY = "UNIQ_ALT_READ_COUNT";
+
+    @Override
+    public List<String> getKeyNames() {
+        return Arrays.asList(UNIQUE_ALT_READ_SET_COUNT_KEY);
+    }
+
+    @Override
+    public List<VCFFormatHeaderLine> getDescriptions() {
+        return Arrays.asList( new VCFFormatHeaderLine(UNIQUE_ALT_READ_SET_COUNT_KEY, 1, VCFHeaderLineType.Integer,
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+                "Number of ALT reads with unique start and mate end positions at a variant site"));
+    }
+
+    @Override
+    public void annotate(final ReferenceContext ref,
+                         final VariantContext vc,
+                         final Genotype g,
+                         final GenotypeBuilder gb,
+                         final ReadLikelihoods<Allele> likelihoods) {
+        if (g.isHomRef()){
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: yup

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+            // skip the normal sample
+            return;
+        }
+
+        final Allele altAllele = vc.getAlternateAllele(0); // assume single-allelic
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: We just take whichever ALT allele is at index 0. It's not ideal but I think that's OK for now. In the future I should add a utility function that lets you pick the alt allele with highest tumor lod.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        final String tumorSampleName = g.getSampleName();
+        Collection<ReadLikelihoods<Allele>.BestAllele> tumorBestAlleles = likelihoods.bestAlleles(tumorSampleName);
+        Map<ImmutablePair<Integer, Integer>, Long> duplicateReadMap = tumorBestAlleles.stream()
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+                .filter(ba -> ba.allele.equals(altAllele) && ba.isInformative())
+                .map(ba -> new ImmutablePair<>(ba.read.getStart(), ba.read.getFragmentLength()))
+                .collect(Collectors.groupingBy(x -> x, Collectors.counting()));
+
+        gb.attribute(UNIQUE_ALT_READ_SET_COUNT_KEY, duplicateReadMap.size());
+    }
+
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: yup

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+import org.broadinstitute.hellbender.utils.read.ArtificialReadUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.util.*;
+
+public class DuplicateReadCountsUnitTest {
+    final String sampleName = "Mark";
```
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java]: negative

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+    final long variantSite = 105;
+
+    final SampleList sampleList = new IndexedSampleList(sampleName);
+
+    final List<Allele> alleles = Arrays.asList(Allele.create((byte) 'C', true), Allele.create((byte) 'A', false));
+    final AlleleList<Allele> alleleList = new IndexedAlleleList<>(alleles);
+    final VariantContext vc = new VariantContextBuilder("source", Integer.toString(chromosomeIndex), variantSite, variantSite, alleles).make();
+
+    @Test
+    public void testSingleDuplicates() throws IOException {
```
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+        final int uniqueReadSetCount = GATKProtectedVariantContextUtils.getAttributeAsInt(genotype, UniqueAltReadCount.UNIQUE_ALT_READ_SET_COUNT_KEY, -1);
+        Assert.assertEquals(uniqueReadSetCount, 1);
+    }
+
+    @Test
+    public void testMultipleDuplicateSets() throws IOException {
+        final UniqueAltReadCount duplicateReadCountsAnnotation = new UniqueAltReadCount();
+
+        // should get three unique sets of ALT reads
+        final int shiftBy1 = 3;
```
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java]: good point. Renamed it to `numUniqueStarts1` and `numUniqueStarts2`

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+        duplicateReadCountsAnnotation.annotate(null, vc, genotypeBuilder1.make(), genotypeBuilder1, likelihoods1);
+
+        final Genotype genotype1 = genotypeBuilder1.make();
+
+        final int uniqueReadSetCount1 = GATKProtectedVariantContextUtils.getAttributeAsInt(genotype1, UniqueAltReadCount.UNIQUE_ALT_READ_SET_COUNT_KEY, -1);
+
+        Assert.assertEquals(uniqueReadSetCount1, shiftBy1);
+
+        // here ALT reads are all distinct
+        final int shiftBy2 = 12;
```
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java]: good call

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+        duplicateReadCountsAnnotation.annotate(null, vc, genotypeBuilder2.make(), genotypeBuilder2, likelihoods2);
+
+        final Genotype genotype2 = genotypeBuilder2.make();
+
+        final int uniqueReadSetCount2 = GATKProtectedVariantContextUtils.getAttributeAsInt(genotype2, UniqueAltReadCount.UNIQUE_ALT_READ_SET_COUNT_KEY, -1);
+
+        Assert.assertEquals(uniqueReadSetCount2, shiftBy2);
+    }
+
+
```
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+    private ReadLikelihoods<Allele> createTestLikelihoods(final boolean shiftReads, Optional<Integer> shiftBy) {
```
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java]: yupp

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java (hunk 1 above)
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/DuplicateReadCountsUnitTest.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/FilterMutectCalls.java [hunk 22]
```diff
@@ -50,6 +50,7 @@ public void onTraversalStart() {
         headerLines.add(new VCFFilterHeaderLine(Mutect2FilteringEngine.MEDIAN_FRAGMENT_LENGTH_DIFFERENCE_FILTER_NAME, "abs(ref - alt) median fragment length"));
         headerLines.add(new VCFFilterHeaderLine(Mutect2FilteringEngine.READ_POSITION_FILTER_NAME, "median distance of alt variants from end of reads"));
         headerLines.add(new VCFFilterHeaderLine(Mutect2FilteringEngine.CONTAMINATION_FILTER_NAME, "contamination"));
+        headerLines.add(new VCFFilterHeaderLine(Mutect2FilteringEngine.DUPLICATED_EVIDENCE_FILTER_NAME, "alt reads are mostly duplicate reads"));
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/FilterMutectCalls.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java (hunk 2 above)
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java (hunk 3 above)
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: which comma?

**Comment by takutosato** [comment]
@fleharty @pshapiro4broad thanks for reviewing. I'm going to migrate the branch to gatk public and make a new PR there.

**Comment by takutosato** [comment]
migrated to broadinstitute/gatk#3134

**Pull request closed without merging**